#  Petite Pandas Data Analysis using Pandas and NumPy
> A series of lessons based on the utilization of Pandas and NumPy to read and create interesting things with files and data. 
- toc: true

# Predictive Analysis
Predictive analysis is the use of statistical, data mining, and machine learning techniques to analyze current and historical data in order to make predictions about future events or behaviors. It involves identifying patterns and trends in data, and then using that information to forecast what is likely to happen in the future.

Predictive analysis is used in a wide range of applications, from forecasting sales and demand, to predicting customer behavior, to detecting fraudulent transactions. It involves collecting and analyzing data from a variety of sources, including historical data, customer data, financial data, and social media data, among others.

The process of predictive analysis typically involves the following steps:
1. Defining the problem and identifying the relevant data sources
2. Collecting and organizing data
3. Exploring and analyzing the data to identify patterns and trends
4. Selecting an appropriate model or algorithm to use for predictions
5. Training and validating the model using historical data
6. Using the model to make predictions on new data
7. Monitoring and evaluating the performance of the model over time

Predictive analysis can help organizations make more informed decisions, improve efficiency, and gain a competitive advantage by leveraging insights from data.

It is most commonly used in retail, where workers try to predict which products would be most popular and try to advertise those products as much as possible, and also Healthcare, where algorithms analyze patterns and reveal prerequisites for diseases and suggest preventive treatment, predict the results of various treatments and choose the best option for each patient individually, and predict disease outbreaks and epidemics.

# 1. Intro to NumPy and the features it consists

Numpy, by definition, is the fundamental package for scientific computing in Python which can be used to perform mathematical operations, provide multidimensional array objects, and makes data analysis much easier. Numpy is very important and useful when it comes to data analysis, as it can easily use its features to complete and perform any mathematical operation, as well as analyze data files.

If you don't already have numpy installed, you can do so using ```conda install numpy``` or ```pip install numpy```

Once that is complete, to import numpy in your code, all you must do is:

In [2]:
import numpy as np

# 2. Using NumPy to create arrays
An array is the central data structure of the NumPy library. They are used as containers which are able to store more than one item at the same time. Using the function ```np.array``` is used to create an array, in which you can create multidimensional arrays. 

Shown below is how to create a 1D array:

In [3]:
a = np.array([1, 2, 3])
print(a) 
# this creates a 1D array

[1 2 3]


How could you create a 3D array based on knowing how to make a 1D array?

In [10]:
# create 3D array here
a = np.array([[1, 2, 3] , [4, 5, 6] , [7, 8, 9]])
print(a)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


Arrays can be printed in different ways, especially a more readable format. As we have seen, arrays are printed in rows and columns, but we can change that by using the ```reshape``` function 

In [11]:
c = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(c.reshape(1, 9)) # organizes it all in a single line of output

[[1 2 3 4 5 6 7 8 9]]


In the code segment below, we can also specially select certain rows and columns from the array to further analyze selective data.

In [12]:
print(c[1:, :2])
# the 1: means "start at row 1 and select all the remaining rows"
# the :2 means "select the first two columns"

[[4 5]
 [7 8]]


# 3. Basic array operations

One of the most basic operations that can be performed on arrays is arithmetic operations. With numpy, it is very easy to perform arithmetic operations on arrays. You can add, subtract, miltiply and divide arrays, just like you would with regular numbers. When performing these operations, numpy applies the operation element-wise, meaning that it performs the operation on each element in the array separately. This makes it easy to perform operations on large amounts of data quickly and efficiently.

In [13]:
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])
print(a + b) # adds each value based on the column the integer is in
print(a - b) # subtracts each value based on the column the integer is in
print(a * b) # multiplies each value based on the column the integer is in
print(a / b) # divides each value based on the column the integer is in

[5 7 9]
[-3 -3 -3]
[ 4 10 18]
[0.25 0.4  0.5 ]


In [14]:
d = np.exp(b)
e = np.sqrt(b)
print(d)
print(e)

[ 54.59815003 148.4131591  403.42879349]
[2.         2.23606798 2.44948974]


From the knowledge of how to use more advanced mathematical expressions than the basic 4 mathematical operations such as exponent and square root, now can you code how to calculate the 3 main trig expressions (sin, cos, tan), natural log, and log10 of a 1D array.

In [18]:
s = np.array([.25, .5, .75])
# calculate sin
print(np.sin(a))
# calculate cos
print(np.cos(a))
# calculate tan
print(np.tan(a))
# calculate natural log
print(np.log(a))
# calculate log10
print(np.log10(a))

[0.84147098 0.90929743 0.14112001]
[ 0.54030231 -0.41614684 -0.9899925 ]
[ 1.55740772 -2.18503986 -0.14254654]
[0.         0.69314718 1.09861229]
[0.         0.30103    0.47712125]


# 4. Data analysis using numpy
Numpy provides a convenient and powerful way to perform data analysis tasks on large data sets. One of the most common tasks in data analysis is finding the mean, median, and standard deviation of a dataset. Numpy provides functions to perform these operations quickly and easily. The mean function calculates the average value of the data, while the median function calculates the middle value in the data. The standard deviation function calculates how spread out the data is from the mean. Additionally, numpy provides functions to find the minimum and maximum values in the data. These functions are very useful for gaining insight into the properties of large datasets and can be used for a wide range of data analysis tasks.

In [19]:
data = np.array([2, 5, 12, 13, 19])
print(np.mean(data)) # finds the mean of the dataset
print(np.median(data)) # finds the median of the dataset
print(np.std(data)) # finds the standard deviation of the dataset
print(np.min(data)) # finds the min of the dataset
print(np.max(data)) # finds the max of the dataset

10.2
12.0
6.04648658313239
2
19


Now from learning this, can you find a different way from how we can solve the sum or products of a dataset other than how we learned before?

In [2]:
import pandas as pd

num1 = int(input("Entrar numbero uno"))
num2 = int(input("Entrar numbero dos"))

df = pd.DataFrame({'numberos': [num1, num2]})

# calc sum
result = df['numberos'].sum()

# result
print(f"The sum of {num1} and {num2} is {result}")

The sum of 1 and 2 is 3


Numpy also has the ability to handle CSV files, which are commonly used to store and exchange large datasets. By importing CSV files into numpy arrays, we can easily perform complex operations and analysis on the data, making numpy an essential tool for data scientists and researchers.

```genfromtxt``` and ```loadtxt``` are two functions in the numpy library that can be used to read data from text files, including CSV files.

```genfromtxt``` is a more advanced function that can be used to read text files that have more complex structures, including CSV files. ```genfromtxt``` can handle files that have missing or invalid data, or files that have columns of different data types. It can also be used to skip header lines or to read only specific columns from the file. 

In [21]:
import numpy as np

padres = np.genfromtxt('files/padres.csv', delimiter=',', dtype=str, encoding='utf-8')
# delimiter indicates that the data is separated into columns which is distinguished by commas
# genfromtxt is used to read the csv file itself
# dtype is used to have numpy automatically detect the data type in the csv file

print(padres)

[['Name' ' Position' ' Average' ' HR' ' RBI' ' OPS' ' JerseyNumber']
 ['Manny Machado' ' 3B' ' .298' ' 32' ' 102' ' .897' ' 13']
 ['Fernando Tatis Jr' ' RF' ' .281' ' 42' ' 97' ' .975' ' 23']
 ['Juan Soto' ' LF' ' .242' ' 27' ' 62' ' .853' ' 22']
 ['Xander Bogaerts' ' SS' ' .307' ' 15' ' 73' ' .833' ' 2']
 ['Nelson Cruz' ' DH' ' .234' ' 10' ' 64' ' .651' ' 32']
 ['Matt Carpenter' ' DH' ' .305' ' 15' ' 37' ' 1.138' ' 14']
 ['Jake Cronenworth' ' 1B' ' .239' ' 17' ' 88' ' .722' ' 9']
 ['Ha-Seong Kim' ' 2B' ' .251' ' 11' ' 59' ' .708' ' 7']
 ['Trent Grisham' ' CF' ' .184' ' 17' ' 53' ' .626' ' 1']
 ['Luis Campusano' ' C' ' .250' ' 1' ' 5' ' .593' ' 12']
 ['Austin Nola' ' C' ' .251' ' 4' ' 40' ' .649' ' 26']
 ['Jose Azocar' ' OF' ' .257' ' 0' ' 10' ' .630' ' 28']]


```loadtxt``` is a simpler function that can be used to read simple text files that have a regular structure, such as files that have only one type of data (such as all integers or all floats). ```loadtxt``` can be faster than ```genfromtxt``` because it assumes that the data in the file is well-structured and can be easily parsed.

In [22]:
import numpy as np

padres = np.loadtxt('files/padres.csv', delimiter=',', dtype=str, encoding='utf-8')
print(padres)

[['Name' ' Position' ' Average' ' HR' ' RBI' ' OPS' ' JerseyNumber']
 ['Manny Machado' ' 3B' ' .298' ' 32' ' 102' ' .897' ' 13']
 ['Fernando Tatis Jr' ' RF' ' .281' ' 42' ' 97' ' .975' ' 23']
 ['Juan Soto' ' LF' ' .242' ' 27' ' 62' ' .853' ' 22']
 ['Xander Bogaerts' ' SS' ' .307' ' 15' ' 73' ' .833' ' 2']
 ['Nelson Cruz' ' DH' ' .234' ' 10' ' 64' ' .651' ' 32']
 ['Matt Carpenter' ' DH' ' .305' ' 15' ' 37' ' 1.138' ' 14']
 ['Jake Cronenworth' ' 1B' ' .239' ' 17' ' 88' ' .722' ' 9']
 ['Ha-Seong Kim' ' 2B' ' .251' ' 11' ' 59' ' .708' ' 7']
 ['Trent Grisham' ' CF' ' .184' ' 17' ' 53' ' .626' ' 1']
 ['Luis Campusano' ' C' ' .250' ' 1' ' 5' ' .593' ' 12']
 ['Austin Nola' ' C' ' .251' ' 4' ' 40' ' .649' ' 26']
 ['Jose Azocar' ' OF' ' .257' ' 0' ' 10' ' .630' ' 28']]


In [ ]:
for i in padres:
    print(",".join(i))

# Pandas
### What is Pandas
Pandas is a Python library used for working with data sets. A python library is something  It has functions for analyzing, cleaning, exploring, and manipulating data.

### Why Use Pandas?
Pandas allows us to analyze big data and make conclusions based on statistical theories. Pandas can clean messy data sets, and make them readable and relevant. Also it is a part of data analysis, and data manipulation.

### What Can Pandas Do?
Pandas gives you answers about the data. Like:
- Is there a correlation between two or more columns?
- What is average value
- Max value
- Min value
- How to load data 
- Delete data 
- Sort Data.

Pandas are also able to delete rows that are not relevant, or contains wrong values, like empty or NULL values. This is called cleaning the data.

# Basics of Pandas.

In [23]:
import pandas as pd
# What this does is it calls the python pandas library and this code segment is needed whenever incorporating pandas.

#### DICTIONARIES AND DATASETS
- One way you are able to manipulate a pandas data set is by creating a dictionary and calling it as seen with the dict data 1 and pd.dataframe which is a way to print the set.

In [24]:
import pandas as pd

data1 = {
  'teams': ["BARCA", "REAL", "ATLETICO"],
  'standings': [1, 2, 3]
}

myvar = pd.DataFrame(data1)

print(myvar)


      teams  standings
0     BARCA          1
1      REAL          2
2  ATLETICO          3


### Indexing and manipulaton of data through lists.
- With pandas you can also organize the data which is one of its biggest perks, we call this indexing, this is when we define the first column in a data frame.

In [25]:
# Here is an example using lists and an index.
import pandas as pd 

score = [5/5, 5/5, 1/5]

myvar = pd.Series(score, index = ["math", "science", "pe"])

print(myvar)

math       1.0
science    1.0
pe         0.2
dtype: float64


# Pandas Classes 
Within pandas the library consits of a lot of functions which allow you to manipulate datasets in lists dictionsaries and csv files here are some of the ones we are going to cover (hint: take notes on these)
- Series
    - Series of indexes, list, dict, etc
- Index
    - A specific element in a list
- PeriodIndex
    - Repeat within a certain index and ability to sort
- DataframeGroupedBy
    - Group seperate data within columns
- Categorical
    - Sets a category for items (good for repeats and organzation)
- Time Stamp
    - Display a specified time


#  PeriodIndex 
- This allows for a way to repeat data over time that it occurs as seen from january 2022 to december 2023. You can use Y for years, M for months, and D for days.


In [28]:
import pandas as pd


time = pd.period_range('2022-06', '2023-07', freq='M')


print(time)

PeriodIndex(['2022-06', '2022-07', '2022-08', '2022-09', '2022-10', '2022-11',
             '2022-12', '2023-01', '2023-02', '2023-03', '2023-04', '2023-05',
             '2023-06', '2023-07'],
            dtype='period[M]')


Now implement a way to show a period index from June 2022 to July 2023 in days.

In [ ]:
# use period index to show - in days - June 2022 to 2023

# Dataframe Grouped By 
- This allows for you to organize your data and calculate the different functions such as
- count(): returns the number of non-null values in each group.
- sum(): returns the sum of values in each group.
- mean(): returns the mean of values in each group.
- min(): returns the minimum value in each group.
- max(): returns the maximum value in each group.
- median(): returns the median of values in each group.
- var(): returns the variance of values in each group.
- agg(): applies one or more functions to each group and returns a new DataFrame with the results.


In [31]:
import pandas as pd

data = {
    'Category': ['E', 'F', 'E', 'F', 'E', 'F', 'E', 'F'],
    'Value': [100, 250, 156, 255, 240, 303, 253, 3014]
}
df = pd.DataFrame(data)


grouped = df.groupby('Category').sum #GUESS WHAT THIS WOULD BE IF WE WERE LOOKING FOR COMBINED TOTALS!()

print(grouped)


<bound method GroupBy.sum of <pandas.core.groupby.generic.DataFrameGroupBy object at 0x7ff84e0f2ce0>>


### Categorical 
- This sets up a category for something and puts it within the categories and allows for better orginzation 

In [32]:
import pandas as pd

colors = pd.Categorical(['yellow', 'orange', 'blue', 'yellow', 'orange'], categories=['yellow', 'orange', 'blue'])

print(colors)


['yellow', 'orange', 'blue', 'yellow', 'orange']
Categories (3, object): ['yellow', 'orange', 'blue']


### Timestamp Class
- This allows to display a single time which can be useful when working with datasets that deal with time allowing you to manipulate the time you do something and how you do it. 

In [33]:
import pandas as pd


timing = pd.Timestamp('2023-02-05 02:00:00')

print(timing)


2023-02-05 02:00:00


# CSV FILES!
- A csv file contains data and within pandas you are able to call the function and you are able to manipulate the data with the certain data classes talked about above. 

- Name, Position, Average, HR, RBI, OPS, JerseyNumber
- Manny Machado, 3B, .298, 32, 102, .897, 13
- Tatis Jr, RF, .281, 42, 97, .975, 23
- Juan Soto, LF, .242, 27, 62, .853, 22
- Xanger Bogaerts, SS, .307, 15, 73, .833, 2
- Nelson Cruz, DH, .234, 10, 64, .651, 32
- Matt Carpenter, DH, .305, 15, 37, 1.138, 14
- Cronezone, 1B, .239, 17, 88, .722, 9
- Ha-Seong Kim, 2B, .251, 11, 59, .708, 7
- Trent Grisham, CF, .184, 17, 53, .626, 1
- Luis Campusano, C, .250, 1, 5, .593, 12
- Austin Nola, C, .251, 4, 40, .649, 26
- Jose Azocar, OF, .257, 0, 10, .630, 28

QUESTION: WHAT DO YOU GUYS THINK THE INDEX FOR THIS WOULD BE?

Can you explain what is going on in this code segment below. (hint: define what ascending= false means, and df. head means)

In [34]:
import pandas as pd

#read csv and sort 'Duration' largest to smallest
df = pd.read_csv('files/padres.csv').sort_values(by=['Name'], ascending=False)

print("--Duration Top 10---------")
print(df.head(10))

print("--Duration Bottom 10------")
print(df.tail(10))
print(', '.join(df.tail(10)))

--Duration Top 10---------
                Name  Position   Average   HR   RBI    OPS   JerseyNumber
3    Xander Bogaerts        SS     0.307   15    73  0.833              2
8      Trent Grisham        CF     0.184   17    53  0.626              1
4        Nelson Cruz        DH     0.234   10    64  0.651             32
5     Matt Carpenter        DH     0.305   15    37  1.138             14
0      Manny Machado        3B     0.298   32   102  0.897             13
9     Luis Campusano         C     0.250    1     5  0.593             12
2          Juan Soto        LF     0.242   27    62  0.853             22
11       Jose Azocar        OF     0.257    0    10  0.630             28
6   Jake Cronenworth        1B     0.239   17    88  0.722              9
7       Ha-Seong Kim        2B     0.251   11    59  0.708              7
--Duration Bottom 10------
                 Name  Position   Average   HR   RBI    OPS   JerseyNumber
4         Nelson Cruz        DH     0.234   10    64  0.6

In [35]:
import pandas as pd


df = pd.read_csv("./files/housing.csv")


mode_total_rooms = df['total_rooms'].mode()


print(f"The mode of the 'total_rooms' column is: {mode_total_rooms}")

The mode of the 'total_rooms' column is: 0    1527.0
Name: total_rooms, dtype: float64


In [36]:
import pandas as pd

df = pd.read_csv("./files/housing.csv")


grouped_df = df.groupby('total_rooms')


agg_df = grouped_df.agg({'total_rooms': 'sum', 'population': 'mean', 'longitude': 'count'})

# WHAT DO YOU GUYS THINK df.agg means in context of pandas and what does it stand for.


print(agg_df)


             total_rooms  population  longitude
total_rooms                                    
2.0                  2.0         6.0          1
6.0                  6.0         8.0          1
8.0                  8.0        13.0          1
11.0                11.0        24.0          1
12.0                12.0        18.0          1
...                  ...         ...        ...
30450.0          30450.0      9419.0          1
32054.0          32054.0     15507.0          1
32627.0          32627.0     28566.0          1
37937.0          37937.0     16122.0          1
39320.0          39320.0     16305.0          1

[5926 rows x 3 columns]


# Our Frontend Data Analysis Project
[Link](https://paravsalaniwal.github.io/T3Project/DataAnalysisProject/)

# Popcorn Hacks
- Complete fill in the blanks for Predictive Analysis Numpy
- Takes notes on Panda where it asks you to
- Complete code segment tasks in Panda and Numpy

# Main Hack
- Make a data file - content is up to you, just make sure there are integer values - and print
- Run Panda and Numpy commands
    - Panda:
        - Find Min and Max values
        - Sort in order - can be order of least to greatest or vice versa
        - Create a smaller dataframe and merge it with your data file
    - Numpy:
        -  Random number generation
        - create a multi-dimensional array (multiple elements)
        - create an array with linearly spaced intervals between values


In [12]:
import pandas as pd
import random

first_names = ['Orlando', 'Soham', 'Colin', 'James', 'Ethan', 'Emily', 'Harvard', 'Joe', 'Sleepy', 'Sleepy Joe', 'Mary', 'Your Mom', 'Collin', 'Max', 'Bartholomew', 'Joel', 'Ava', 'Oliver', 'Charlotte', 'Lucas', 'Harper', 'Mason', 'Amelia', 'Evelyn', 'Logan', 'Ella', 'Joseph', 'Liam', 'Connor', 'Abby', 'Austin', 'Avery', 'Bryce', 'Camila', 'Carter', 'Charlotte', 'Eli', 'Ella', 'Emma', 'Grace', 'Hazel', 'Isaac', 'Jade', 'Jaxon', 'Jocelyn', 'Levi', 'Lila', 'Liliana', 'Mia', 'Mila', 'Natalie', 'Oliver', 'Piper', 'Sebastian', 'Jillian',]
last_names = ['Smith', 'Johnson', 'Williams', 'Brown', 'Jones', 'Carcamo', 'Weis', 'Kamat', 'Joe', 'Biden', 'Ward', 'Campbell', 'Gonzalez', 'Bailey', 'Bennett', 'Wong', 'Perry', 'Ross', 'Dixon', 'Morgan', 'Owens', 'Harris', 'Lopez', 'Foster', 'Chambers', 'Hayes', 'Porter', 'Peters', 'Sullivan', 'Wright', 'Edwards', 'Mitchell', 'Esteban', 'Cox', 'Hill', 'Harper', 'Ceja', 'Jr', 'The III', 'The II', 'The IIII', 'The IIIII']
professions = ['Doctor', 'Heart Seller', 'Professional Coach', 'Lifeguard', 'Astronaut', 'Robotics Engineer', 'Artificial Intelligence Developer', 'Magician', 'Firefighter', 'Astrologer', 'Barista', 'Psychiatrist', 'Crime Scene Investigator', 'Private Investigator', 'Dog Walker', 'Chocolatier', 'Cartoonist', 'Meteorologist', 'Circus Performer', 'Graphic Designer', 'Fashion Designer', 'Musician', 'Professional Gamer', 'Baker', 'Stand-up Comedian', 'Movie Director', 'Gardener', 'Event Planner', 'Archaeologist', 'Translator', 'Social Media Influencer', 'Fitness Trainer', 'Mechanic', 'Chef', 'Hair Stylist', 'Mortician', 'Wine Sommelier', 'Surgeon', 'Yoga Instructor', 'Author', 'Interior Designer', 'Tattoo Artist', 'Pilot', 'Magistrate', 'Historian', 'Art Dealer', 'Cartographer', 'Acrobat', 'Paranormal Investigator', 'Professional Border Hopper', 'Taco Muncher', 'Horchata Slurper', 'Traditional Curry Chomper', 'Professional "Mom" investigator', 'School Shooter', 'Obesity Lord', ]

# Create a DataFrame with randomized first names, last names, hours of work, and professions
data = []
for i in range(100):
    data.append({
        'first_name': random.choice(first_names),
        'last_name': random.choice(last_names),
        'age': random.randint(1, 120),
        'salary': random.randint(-20000, 250000),
        'profession': random.choice(professions)
    })
df = pd.DataFrame(data)

print(df)

df.to_csv('goofy.csv', index=False)

   first_name  last_name  age  salary                  profession
0     Orlando     Morgan   26  129306  Professional Border Hopper
1      Evelyn       Weis   25  233457               Archaeologist
2     Natalie     Harper   95  145172                Obesity Lord
3     Harvard     Porter   55    6619                Hair Stylist
4      Carter    Esteban   93  131439   Traditional Curry Chomper
..        ...        ...  ...     ...                         ...
95       Jade        Cox   22  182905    Crime Scene Investigator
96       Abby   The IIII   29  131219                     Surgeon
97        Max  The IIIII   88  115416                Hair Stylist
98     Sleepy   Gonzalez   85   20450                        Chef
99       Liam    The III   49  198129          Professional Gamer

[100 rows x 5 columns]


In [15]:
import pandas as pd

df = pd.read_csv('goofy.csv')

max_age = df['age'].max()
min_age = df['age'].min()
avg_age = df['age'].mean()

max_salary = df['salary'].max()
min_salary = df['salary'].min()
avg_salary = df['salary'].mean()

most_common_profession = df['profession'].mode()[0]
most_common_fn = df['first_name'].mode()[0]
most_common_ln = df['last_name'].mode()[0]

print(f"Max age - {max_age}")
print(f"Min age - {min_age}")
print(f"Average age - {avg_age:.2f}")
print(f"Max salary - {max_salary}")
print(f"Min salary - {min_salary}")
print(f"Average salary - {avg_salary:.2f}")
print(f"Most common profession - {most_common_profession}")
print(f"Most common first name - {most_common_fn}")
print(f"Most common last name - {most_common_ln}")

Max age - 120
Min age - 2
Average age - 64.42
Max salary - 247410
Min salary - -16787
Average salary - 107696.13
Most common profession - Interior Designer
Most common first name - Emily
Most common last name - Mitchell


In [19]:
import numpy as np
import random

random_list = [random.randint(1, 50) for i in range(10)]

array = np.array(random_list)

print(array)

[43 36 46 47 46 25 33 44 37 30]


In [20]:
max_num = np.amax(array)

square = int(np.sqrt(max_num)) ** 2

triple_d_array = np.zeros((square, square, square))

print(triple_d_array)

[[[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 [[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 [[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 ...

 [[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 [[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 [[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]]


In [21]:
linear_array = np.linspace(0, 1, num=square ** 3).reshape(square, square, square)

print(linear_array)

[[[0.00000000e+00 2.14339299e-05 4.28678598e-05 ... 7.07319687e-04
   7.28753617e-04 7.50187547e-04]
  [7.71621477e-04 7.93055407e-04 8.14489337e-04 ... 1.47894116e-03
   1.50037509e-03 1.52180902e-03]
  [1.54324295e-03 1.56467688e-03 1.58611081e-03 ... 2.25056264e-03
   2.27199657e-03 2.29343050e-03]
  ...
  [2.54635087e-02 2.54849427e-02 2.55063766e-02 ... 2.61708284e-02
   2.61922624e-02 2.62136963e-02]
  [2.62351302e-02 2.62565641e-02 2.62779981e-02 ... 2.69424499e-02
   2.69638838e-02 2.69853178e-02]
  [2.70067517e-02 2.70281856e-02 2.70496195e-02 ... 2.77140714e-02
   2.77355053e-02 2.77569392e-02]]

 [[2.77783732e-02 2.77998071e-02 2.78212410e-02 ... 2.84856929e-02
   2.85071268e-02 2.85285607e-02]
  [2.85499946e-02 2.85714286e-02 2.85928625e-02 ... 2.92573143e-02
   2.92787483e-02 2.93001822e-02]
  [2.93216161e-02 2.93430500e-02 2.93644840e-02 ... 3.00289358e-02
   3.00503697e-02 3.00718037e-02]
  ...
  [5.32418819e-02 5.32633158e-02 5.32847498e-02 ... 5.39492016e-02
   5.39706

# Grading
The grading will be binary - all or nothing; no partial credit
- 0.3 for all the popcorn hacks
> - done-erino
- 0.6 for the main hack - CSV file
> - Picasso is finished
- 0.1 for going above and beyond in the main hack
> - I mean look at the picasso